File: Session_01_AlgorithmAnalysis/multi_language_implementations/complexity_examples.java
import java.util.Scanner;

public class complexity_examples {
    public static void demonstrateConstantTime(int n) {
        int operations = 0;
        int a = 10;
        operations++;
        int b = 20;
        operations++;
        int sum = a + b;
        operations++;
        System.out.println("O(1) Constant Time:");
        System.out.println("  Input size (n): " + n);
        System.out.println("  Actual operations counted: " + operations + "\n");
    }

    public static void demonstrateLogarithmicTime(int n) {
        int operations = 0;
        int temp = n;
        System.out.println("O(log n) Logarithmic Time:");
        System.out.println("  Dividing input size " + n + " by 2 repeatedly:");
        while (temp > 1) {
            System.out.println("    Current value: " + temp);
            temp /= 2;
            operations++;
        }
        System.out.println("  Input size (n): " + n);
        System.out.println("  Actual operations counted: " + operations + "\n");
    }

    public static void demonstrateLinearTime(int n) {
        int operations = 0;
        for (int i = 0; i < n; i++) {
            operations++;
        }
        System.out.println("O(n) Linear Time:");
        System.out.println("  Input size (n): " + n);
        System.out.println("  Actual operations counted: " + operations + "\n");
    }

    public static void demonstrateLinearithmicTime(int n) {
        int operations = 0;
        for (int i = 0; i < n; i++) {
            int temp = n;
            while (temp > 1) {
                temp /= 2;
                operations++;
            }
        }
        System.out.println("O(n log n) Linearithmic Time:");
        System.out.println("  Input size (n): " + n);
        System.out.println("  Actual operations counted: " + operations + "\n");
    }

    public static void demonstrateQuadraticTime(int n) {
        int operations = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                operations++;
            }
        }
        System.out.println("O(n^2) Quadratic Time:");
        System.out.println("  Input size (n): " + n);
        System.out.println("  Actual operations counted: " + operations + " (n * n)\n");
    }

    public static void main(String[] args) {
        int n = 16;
        System.out.println("=== Time Complexity Visualizer ===");
        System.out.println("This program counts operations to demonstrate theoretical time complexities.\n");
        demonstrateConstantTime(n);
        demonstrateLogarithmicTime(n);
        demonstrateLinearTime(n);
        demonstrateLinearithmicTime(n);
        demonstrateQuadraticTime(n);
        System.out.print("Press Enter to exit...");
        Scanner scanner = new Scanner(System.in);
        scanner.nextLine();
        scanner.close();
    }
}

File: Session_01_AlgorithmAnalysis/multi_language_implementations/complexity_examples.kt
import java.util.Scanner

fun demonstrateConstantTime(n: Int) {
    var operations = 0
    val a = 10
    operations++
    val b = 20
    operations++
    val sum = a + b
    operations++
    println("O(1) Constant Time:")
    println("  Input size (n): $n")
    println("  Actual operations counted: $operations\n")
}

fun demonstrateLogarithmicTime(n: Int) {
    var operations = 0
    var temp = n
    println("O(log n) Logarithmic Time:")
    println("  Dividing input size $n by 2 repeatedly:")
    while (temp > 1) {
        println("    Current value: $temp")
        temp /= 2
        operations++
    }
    println("  Input size (n): $n")
    println("  Actual operations counted: $operations\n")
}

fun demonstrateLinearTime(n: Int) {
    var operations = 0
    for (i in 0 until n) {
        operations++
    }
    println("O(n) Linear Time:")
    println("  Input size (n): $n")
    println("  Actual operations counted: $operations\n")
}

fun demonstrateLinearithmicTime(n: Int) {
    var operations = 0
    for (i in 0 until n) {
        var temp = n
        while (temp > 1) {
            temp /= 2
            operations++
        }
    }
    println("O(n log n) Linearithmic Time:")
    println("  Input size (n): $n")
    println("  Actual operations counted: $operations\n")
}

fun demonstrateQuadraticTime(n: Int) {
    var operations = 0
    for (i in 0 until n) {
        for (j in 0 until n) {
            operations++
        }
    }
    println("O(n^2) Quadratic Time:")
    println("  Input size (n): $n")
    println("  Actual operations counted: $operations (n * n)\n")
}

fun main() {
    val n = 16
    println("=== Time Complexity Visualizer ===")
    println("This program counts operations to demonstrate theoretical time complexities.\n")
    demonstrateConstantTime(n)
    demonstrateLogarithmicTime(n)
    demonstrateLinearTime(n)
    demonstrateLinearithmicTime(n)
    demonstrateQuadraticTime(n)
    print("Press Enter to exit...")
    Scanner(System.`in`).nextLine()
}
